unistd/sys.c

```diff
@@ -68,88 +94,109 @@ int execv(const char *path, char *const argv[])
 	return execve(path, argv, environ);
 }
 
-#define PATH_DELIM ':'
 
 int execve(const char *file, char *const argv[], char *const envp[])
 {
 	int fd, noargs = 0, err;
-	char *interp = exec_buffer, *end, **sb_args = NULL;
-	char *canonical_path;
+	char *interp, *end;
 	char *path = getenv("PATH");
-	char filepath_buffer[PATH_MAX];
-	int filename_len = strlen(file);
+	int fileNameLen = strlen(file);
 	struct stat buf;
 
 	fflush(NULL);
+	sys_clear();
+
+	sys_common.execBuff = calloc(PATH_MAX, sizeof(char));
+	sys_common.filePathBuff = calloc(PATH_MAX, sizeof(char));
+	if (sys_common.execBuff == NULL || sys_common.filePathBuff == NULL) {
+		sys_clear();
+		return -ENOMEM;
+	}
+
+	interp = sys_common.execBuff;
 
 	if (!strchr(file, '/') && path) {
 		while (*path) {
 			char* delim = strchrnul(path, ':');
 
 			int path_len = delim - path;
-			if (path_len + 1 + filename_len + 1 <= sizeof(filepath_buffer)) {
-				memcpy(filepath_buffer, path, path_len);
-				filepath_buffer[path_len] = '/';
-				memcpy(filepath_buffer + path_len + 1, file, filename_len);
-				filepath_buffer[path_len + 1 + filename_len] = '\0';
+			if (path_len + 1 + fileNameLen + 1 <= PATH_MAX) {
+				memcpy(sys_common.filePathBuff, path, path_len);
+				sys_common.filePathBuff[path_len] = '/';
+				memcpy(sys_common.filePathBuff + path_len + 1, file, fileNameLen);
+				sys_common.filePathBuff[path_len + 1 + fileNameLen] = '\0';
 
 				/* check if file exists with symlink resolving */
-				if ((canonical_path = resolve_path(filepath_buffer, NULL, 1, 0)) != NULL) {
-					free(canonical_path);
-					file = filepath_buffer;
+				if ((sys_common.canonicalPath = resolve_path(sys_common.filePathBuff, NULL, 1, 0)) != NULL) {
+					free(sys_common.canonicalPath);
```

Comment:
To avoid possible double free:
```suggestion
					free(sys_common.canonicalPath);
					sys_common.canonicalPath = NULL;
```

Reply:
Done